README.md
<p align="center">
<img src="docs/nincid.png"><br><br>
A multiplatform Electron solution to <a href="https://github.com/jaburns/NintendoSpy">NintendoSpy</a>. Please see their repository for any and all setup of Arduino devices.<br>
Source code for NinCID is available on <a href="https://github.com/slashinfty/nincid">GitHub</a>.
</p>

## Releases

Current Version: 0.2.3

<a href="https://github.com/slashinfty/nincid/releases/latest/download/NinCID_Setup_0.2.3.exe">Windows (.exe)</a><br>
<a href="https://github.com/slashinfty/nincid/releases/latest/download/NinCID-0.2.3.dmg">Mac (.dmg)</a><br>
<a href="https://github.com/slashinfty/nincid/releases/latest/download/nincid_0.2.3_amd64.deb">Linux (.deb)</a><br>
<a href="https://github.com/slashinfty/nincid/releases/latest/download/NinCID-0.2.3.AppImage">Linux (.AppImage)</a>

Note: when launching the .AppImage, add the option `--no-sandbox`

View the [Changelog](changelog.md).

## Skins
Built-in skins are the minimalist versions of the NES, SNES, N64, and GameCube from [Proximity Sound](https://proximitysound.com/skins/). There is an additional Super Game Boy console skin in the colors of the original Game Boy.

Custom skins require a `skin.json` in the format below with all corresponding images in a single folder (select the folder when loading the custom skin).

```json
{
  "name": "name of skin",
  "author": "name of author",
  "console": "nes",
  "height": 441,
  "width": 981,
  "background": "back.png",
  "buttons": [
    {
      "name": "button name",
      "image": "button.png",
      "x": "200",
      "y": "150",
      "height": "30",
      "width": "30",
      "range": "35"
    },
    {
      "name": "different button",
      "image": "another.png",
      "x": "410",
      "y": "175"
    }
  ]
}
```

Name and author are optional. Console must be `nes`, `snes`, `n64`, `gcn`, or `sgb`. Height and width are mandatory, and should match the background image's dimensions. For buttons, name must be one of the following (corresponding to the button): `a, b, x, y, z, l, r, start, select, up, down, left, right, stick, cup, cdown, cleft, cright, cstick`. Image, x, and y are mandatory. Height and width are optional, but allow you to reuse the same image for multiple buttons of different sizes. Range is mandatory for analog sticks, and not allowed for buttons.

## Cloning
```bash
# Clone repository
git clone https://github.com/slashinfty/nincid.git && cd nincid

# Install
yarn install

# If permissions are not set on chrome-sandbox when attempting to start
sudo chown root node_modules/electron/dist/chrome-sandbox && sudo chmod 4755 node_modules/electron/dist/chrome-sandbox

# If there is an error about bindings and node versions after starting
rm yarn.lock && rm -rf node_modules/serialport node_modules/@serialport
yarn install
./node_modules/.bin/electron-rebuild # For Linux/Mac
# For Windows:
# .\node_modules\.bin\electron-rebuild.cmd 

# Start
yarn start
```


changelog.md
# Changelog

## v0.2.3 (24 June 2020)

Reverting resizability from v0.2.2 thanks to `win.setContentSize()` correctly resizing the window to match the background.

## v0.2.2 (22 June 2020)

Changed `BrowserWindow` resizability from `false` to `true` as a temporary solution to Windows and Mac not resizing the window to match the background.

## v0.2.1 (22 June 2020)

Changed serial port check from `if (port.manufacturer != undefined)` to `if (port.productId != undefined)`. Using `win.setMinimumSize()` workaround to shrink screen on `win.setSize()`. Changed `nincid.svg` from font-based to path-based. Natively building on all platforms.

## v0.2.0 (17 June 2020)

Implemented custom skins.

## v0.1.1 (15 June 2020)

Removed `"skipTaskbar": true` from `BrowserWindow` constructor to eliminate the app pseudo-closing on Windows.

## v0.1.0 (15 June 2020)

Initial release.

main/index.js
const { app, BrowserWindow, shell } = require('electron')
const path = require('path')

const isDevelopment = process.env.NODE_ENV !== 'production'

function createWindow () {
  // Create the browser window.
  const win = new BrowserWindow({
      webPreferences: {
        nodeIntegration: true,
        enableRemoteModule: true,
        preload: path.join(__dirname, 'preload.js')
      },
      title: 'NinCID',
      height: 300,
      width: 650,
      maximizable: false,
      resizable: false,
      backgroundColor: '#000'
  })

  // Remove the menu.
  win.setMenu(null);

  // and load the index.html of the app.
  win.loadFile('./public/index.html')

  // Links open in native browser.
  win.webContents.on('new-window', (event, url) => {
    event.preventDefault()
    shell.openExternal(url)
  })

  // Uncomment next line to enable dev tools.
  // win.webContents.openDevTools()
}

app.allowRendererProcessReuse = false

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.whenReady().then(createWindow)

// Quit when all windows are closed.
app.on('window-all-closed', () => {
  // On macOS it is common for applications and their menu bar
  // to stay active until the user quits explicitly with Cmd + Q
  if (process.platform !== 'darwin') {
    app.quit()
  }
})

app.on('activate', () => {
  // On macOS it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow()
  }
})

public/menu.js
const { remote } = require('electron')
const { Menu, MenuItem } = remote
var showReturnToSetup = false

const menu = new Menu()
menu.append(new MenuItem(
  {
    label: 'Return to Setup',
    visible: showReturnToSetup,
    click(e, win) {
      document.body.style.backgroundImage = "none"
      document.getElementById("settings-wrapper").style.removeProperty("display")
      win.setSize(500, 275)
      document.getElementById("buttons-container").innerHTML = ""
      showReturnToSetup = false
    }
  }
))
menu.append(new MenuItem(
  {
    label: 'Open Developer Tools',
    role: 'toggleDevTools'
  }
))
menu.append(new MenuItem(
  {
    label: 'Quit NinCID',
    role: 'quit'
  }
))

window.addEventListener('contextmenu', e => {
  e.preventDefault()
  menu.popup({ window: remote.getCurrentWindow() })
}, false)

public/settings_interface.js
const SerialPort = require('serialport')

// Disables alternate console color selection
function disableAltConsole() {
  if (document.getElementById('console-select').value === 'n64' || document.getElementById('console-select').value === 'gcn' || document.getElementById('console-select').value === 'sgb') {
    document.getElementById('color-select').options.namedItem('console-alt').disabled = true
    if (document.getElementById('color-select').options.namedItem('console-alt').selected) document.getElementById('color-select').options.namedItem('console').selected = true
  } else {
    document.getElementById('color-select').options.namedItem('console-alt').disabled = false
  }
}

// Scans for new devices
function scanPorts() {
  let PortOptions = new DocumentFragment()
  SerialPort.list().then(ports => {
    ports.forEach(port => {
      if (port.manufacturer != undefined) {
        const PortOptionElement = document.createElement('option')
        PortOptionElement.setAttribute('value', port.path)
        PortOptionElement.innerText = port.path
        PortOptions.appendChild(PortOptionElement)
      }
    })
    document.getElementById("port-select").appendChild(PortOptions)
    if (document.getElementById("port-select").length != 0) {
      document.getElementById("scan-button").style.display = "none"
      document.getElementById("load-buttons").style.display = "block"
    }
  })
}

public/display_interface.js
const upath = require('upath');
const ByteLength = require('@serialport/parser-byte-length')
const { dialog } = require('electron').remote
const fs = require('fs')
var skinPath;

const consoleObject = [
  {
    "name": "nes",
    "length": 9,
    "buttons": [
      {"button": "a", "input": 0},
      {"button": "b", "input": 1},
      {"button": "select", "input": 2},
      {"button": "start", "input": 3},
      {"button": "up", "input": 4},
      {"button": "down", "input": 5},
      {"button": "left", "input": 6},
      {"button": "right", "input": 7}
    ]
  },
  {
    "name": "snes",
    "length": 17,
    "buttons": [
      {"button": "b", "input": 0},
      {"button": "y", "input": 1},
      {"button": "select", "input": 2},
      {"button": "start", "input": 3},
      {"button": "up", "input": 4},
      {"button": "down", "input": 5},
      {"button": "left", "input": 6},
      {"button": "right", "input": 7},
      {"button": "a", "input": 8},
      {"button": "x", "input": 9},
      {"button": "l", "input": 10},
      {"button": "r", "input": 11}
    ]
  },
  {
    "name": "n64",
    "length": 33,
    "buttons": [
      {"button": "a", "input": 0},
      {"button": "b", "input": 1},
      {"button": "z", "input": 2},
      {"button": "start", "input": 3},
      {"button": "up", "input": 4},
      {"button": "down", "input": 5},
      {"button": "left", "input": 6},
      {"button": "right", "input": 7},
      {"button": "l", "input": 10},
      {"button": "r", "input": 11},
      {"button": "cup", "input": 12},
      {"button": "cdown", "input": 13},
      {"button": "cleft", "input": 14},
      {"button": "cright", "input": 15}
    ],
    "sticks": [
      {"stick": "stick", "xinput": 16, "yinput": 24}
    ]
  },
  {
    "name": "gcn",
    "length": 65,
    "buttons": [
      {"button": "start", "input": 3},
      {"button": "y", "input": 4},
      {"button": "x", "input": 5},
      {"button": "b", "input": 6},
      {"button": "a", "input": 7},
      {"button": "l", "input": 9},
      {"button": "r", "input": 10},
      {"button": "z", "input": 11},
      {"button": "up", "input": 12},
      {"button": "down", "input": 13},
      {"button": "right", "input": 14},
      {"button": "left", "input": 15}
    ],
    "sticks": [
      {"stick": "stick", "xinput": 16, "yinput": 24},
      {"stick": "cstick", "xinput": 32, "yinput": 40}
    ]
  },
  {
    "name": "sgb",
    "length": 17,
    "buttons": [
      {"button": "a", "input": 0},
      {"button": "b", "input": 1},
      {"button": "select", "input": 2},
      {"button": "start", "input": 3},
      {"button": "up", "input": 4},
      {"button": "down", "input": 5},
      {"button": "left", "input": 6},
      {"button": "right", "input": 7}
    ]
  }
]

function loadDisplay() {
  // Get settings, find object, load background.
  document.getElementById("settings-wrapper").style.display = "none"
  const consul = document.getElementById("console-select").value
  const color = document.getElementById("color-select").value
  skinPath = upath.toUnix(upath.join(__dirname, "../static/skins/", consul))
  const skinJson = require(`${skinPath}/skin.json`)
  const background = skinJson.background.find(skin => skin.name === color).image
  const backPath = upath.toUnix(upath.join(skinPath, background))
  document.body.style.backgroundImage = "url(" + backPath + ")"
  
  // Resize window based on height/width in object.
  const win = require("electron").remote.BrowserWindow.getFocusedWindow();
  win.setMinimumSize(skinJson.width, skinJson.height)
  win.setContentSize(skinJson.width, skinJson.height)

  // Set buttons in place.
  const ButtonHolder = new DocumentFragment()
  skinJson.buttons.forEach(button => {
    const ButtonElement = document.createElement("img")
    const buttonID = button.name
    ButtonElement.setAttribute("id", buttonID)
    const buttonPath = upath.toUnix(upath.join(skinPath, button.image))
    ButtonElement.setAttribute("src", buttonPath.toString())
    const widthHeight = button.hasOwnProperty("width") ? "height:" + button.height + "px;width:" + button.width + "px;" : ""
    const vis = button.hasOwnProperty("range") ? "visibility:visible;" : "visibility:hidden;"
    ButtonElement.setAttribute("style", "position:fixed;left:" + button.x + "px;top:" + button.y + "px;" + vis + widthHeight)
    ButtonHolder.appendChild(ButtonElement)
  })
  document.getElementById("buttons-container").appendChild(ButtonHolder)

  // Open port and start.
  port = new SerialPort(document.getElementById("port-select").value, {
    baudRate: 115200
  });

  readPort(port, consul)
}

function customSkin() {
  // Open dialog to select directory.
  const win = require("electron").remote.getCurrentWindow();
  const dir = dialog.showOpenDialogSync({
    "title": "Select Custom Skin Directory",
    "properties": [
      "openDirectory"
    ]
  })[0]
  const jsonPath = upath.toUnix(upath.join(dir, "skin.json"))
  if (!fs.existsSync(jsonPath)) {
    dialog.showErrorBox("Error Loading Custom Skin", "Can not find " + jsonPath)
  } else {
    try {
      // Get settings and load background.
      document.getElementById("settings-wrapper").style.display = "none"
      let rawSkinData = fs.readFileSync(jsonPath)
      let skinJson = JSON.parse(rawSkinData)
      const allowedConsoles = ["nes", "snes", "n64", "gcn", "sgb"]
      if (!skinJson.hasOwnProperty("console") || !allowedConsoles.includes(skinJson.console)) throw "missing/incorrect console"
      if (!skinJson.hasOwnProperty("background")) throw "missing background"
      const background = upath.toUnix(upath.join(dir, skinJson.background))
      document.body.style.backgroundImage = "url(" + background + ")"

      // Resize window based on height/width in object.
      if (!skinJson.hasOwnProperty("width") || !skinJson.hasOwnProperty("height")) throw "missing height/width"
      win.setMinimumSize(skinJson.width, skinJson.height)
      win.setContentSize(skinJson.width, skinJson.height)

      // Set buttons in place.
      const allowedButtons = ["a", "b", "x", "y", "z", "l", "r", "start", "select", "up", "down", "left", "right", "stick", "cup", "cdown", "cleft", "cright", "cstick"]
      const ButtonHolder = new DocumentFragment()
      skinJson.buttons.forEach(button => {
        if (!allowedButtons.includes(button.name)) throw button.name + " is not a valid button name."
        const ButtonElement = document.createElement("img")
        const buttonID = button.name
        ButtonElement.setAttribute("id", buttonID)
        if (!button.hasOwnProperty("image") || !fs.existsSync(upath.toUnix(upath.join(dir, button.image)))) throw "no image"
        const buttonPath = upath.toUnix(upath.join(dir, button.image))
        ButtonElement.setAttribute("src", buttonPath.toString())
        const widthHeight = button.hasOwnProperty("width") ? "height:" + button.height + "px;width:" + button.width + "px;" : ""
        const vis = button.hasOwnProperty("range") ? "visibility:visible;" : "visibility:hidden;"
        if (!button.hasOwnProperty("x") || !button.hasOwnProperty("y")) throw "missing x/y coordinates for " + button.name
        ButtonElement.setAttribute("style", "position:fixed;left:" + button.x + "px;top:" + button.y + "px;" + vis + widthHeight)
        ButtonHolder.appendChild(ButtonElement)
      })
      document.getElementById("buttons-container").appendChild(ButtonHolder)
  

      // Open port and start.
      let port = new SerialPort(document.getElementById("port-select").value, {
        baudRate: 115200
      });
      readPort(port, skinJson.console)
    } catch (error) {
      // Display the error message
      dialog.showErrorBox("Error Loading Custom Skin", error.toString())
      // Set background to black and put up settings
      document.body.style.backgroundImage = "none"
      document.getElementById("settings-wrapper").style.removeProperty("display")
      win.setMinimumSize(650, 300)
      win.setContentSize(650, 300)
      // Remove all buttons and sticks
      document.getElementById("buttons-container").innerHTML = ""
    }
  }
}

function readPort(activePort, consoleName) {
  // Prepare elements based on console.
  const activeConsole = consoleObject.find(obj => obj.name === consoleName)
  const consoleButtons = activeConsole.buttons
  const consoleSticks = activeConsole.hasOwnProperty('sticks') ? activeConsole.sticks : null
  if (consoleSticks != null) { consoleSticks.forEach(stick => {
    const el = document.getElementById(stick.stick)
    stick.left = parseInt(el.style.left)
    stick.top = parseInt(el.style.top)
    stick.range = parseInt(skinJson.buttons.find(o => o.name === stick.stick).range)
  })}
  
  // Functions for interpreting analog sticks.
  const isHighBit = bit => bit & 0x0F !== 0
  const stickRead = bitArray => {
    let val = 0
    if (consoleName === 'n64') {
      for (let i = 1; i < 8; i++) { if (isHighBit(bitArray[i])) { val |= 1 << (7 - i) } }
      return (isHighBit(bitArray[0]) ? (val - 128) : val) * 0.0078125
    } else {
      for (let i = 0; i < 8; i++) { if (isHighBit(bitArray[i])) { val |= 1 << (7 - i) } }
      return (val - 128) * 0.0078125
    }
  };

  // Read data from the port.
  const parser = activePort.pipe(new ByteLength({ length: activeConsole.length }))
  parser.on('data', data => {
    let offset = data.indexOf(10) + 1
    if (consoleSticks != null) {
      consoleSticks.forEach(stk => {
        let x = stickRead(Array.from({ length: 8 }, (_, i) => data[((stk.xinput + offset) % data.length) + i]))
        let y = stickRead(Array.from({ length: 8 }, (_, i) => data[((stk.yinput + offset) % data.length) + i]))
        let el = document.getElementById(stk.stick)
        el.style.left = ((x * stk.range) + stk.left) + "px"
        el.style.top = (stk.top - (y * stk.range)) + "px"
      })
    }
    consoleButtons.forEach(but => {
      document.getElementById(but.button).style.visibility = data[(but.input + offset) % data.length] === 0 ? "hidden" : "visible"
    })
  })
}